SIP-012: Burn height selection for network-upgrade to introduce new cost-limits (#41)
stacksgov/sips: sips/sip-012/sip-012-cost-limits-network-upgrade.md

(define-read-only (cost_unwrap_ret (n uint))
    (runtime u339))

(define-read-only (cost_unwrap_err_or_ret (n uint))
    (runtime u339))

(define-read-only (cost_is_okay (n uint))
    (runtime u287))

(define-read-only (cost_is_none (n uint))
    (runtime u287))

(define-read-only (cost_is_err (n uint))
    (runtime u287))

(define-read-only (cost_is_some (n uint))
    (runtime u287))

(define-read-only (cost_unwrap (n uint))
    (runtime u287))

(define-read-only (cost_unwrap_err (n uint))
    (runtime u287))

(define-read-only (cost_try_ret (n uint))
    (runtime u287))

(define-read-only (cost_match (n uint))
    (runtime u287))

(define-read-only (cost_or (n uint))
    (runtime (linear n u3 u149)))

(define-read-only (cost_and (n uint))
    (runtime (linear n u3 u149)))

(define-read-only (cost_append (n uint))
    (runtime (linear n u71 u176)))

(define-read-only (cost_concat (n uint))
    (runtime (linear n u75 u244)))

(define-read-only (cost_as_max_len (n uint))
    (runtime u475))

(define-read-only (cost_contract_call (n uint))
    (runtime u153))

(define-read-only (cost_contract_of (n uint))
    (runtime u13400))

(define-read-only (cost_principal_of (n uint))

>     (runtime u235))

Reviewer: ```suggestion
    (runtime u999))
```

fixed an issue with this benchmark for `principal_of` and re-ran it

# Preamble

SIP Number: 012

Title: Burn height selection for network-upgrade to introduce new cost-limits

Authors:
* 0xAsteria <asteria@syvita.org>
* Aaron Blankstein <aaron@hiro.so>
* Diwaker Gupta <diwaker@hiro.so>
* Jason Lau <jason@okcoin.com>
* Ludovic Galabru <ludo@hiro.so>
* Trevor Owens <trevor@stacks.ac>
* Xan Ditkoff <xan@daemontechnologies.co>

Consideration: Governance, Technical

Type: Consensus

Status: Draft

Created: 2021-10-08

License: BSD 2-Clause

Sign-off:

Discussions-To: https://github.com/stacksgov/sips

# Abstract

> The current Clarity cost limits were set very conservatively in Stacks 2.0: transactions with contract-calls frequently exceed these limits, which negatively affects transaction throughput. This SIP proposes an update to these cost-limits via a network upgrade and further, that the network upgrade be executed at a block height chosen by an off-chain process described in this SIP.

Reviewer 2: Resolved in ba4c65a830ae075ab6c0fced748873ff78b26d7d

# License and Copyright

This SIP is made available under the terms of the Creative Commons CC0 1.0 Universal license, available at https://creativecommons.org/publicdomain/zero/1.0/ This SIP’s copyright is held by the Stacks Open Internet Foundation.

# Introduction

> Blocks on the Stacks blockchain tend to have anywhere between 10 to 50 transactions per block -- this is lower than the demands of many workloads and also lower than the theoretical maximum one would expect. On the other hand, the mempool consistently has hundreds of valid transactions pending at any given time; at peak there have been several thousand pending transactions. So what is preventing more transactions from being included in blocks?

Reviewer 3: The number of transactions is not, and has never been a useful measure of system throughput.  What is important, and what this SIP is really about, is changing the amount of computational work that happens per block.  Transactions are merely an application-level frame for representing distinct computations.

> The answer is cost-limits. An analysis of “full blocks” (meaning blocks where at least one cost dimension exceeds the block limit -- see SIP-006 for a description of all cost dimensions) found that roughly 72% of those blocks hit the `runtime` limit and ~27% hit the `read_count` limit. Another analysis of all transactions rejected due to cost limits found that 90% of those transactions exceeded the `runtime` limit. So the `runtime` limits are the primary bottleneck right now.

Reviewer 3: Can you add the spreadsheet and graph as a supplemental item to this SIP?  See SIP-000 for details on how to do this.

In the last few months, the [clarity-benchmarking](https://github.com/blockstack/clarity-benchmarking) project has done rigorous benchmarking on contemporary hardware to come up with more accurate cost-limits, with a focus on the `runtime` limits. The updated cost-limits are described in detail in [this forum post](https://forum.stacks.org/t/more-accurate-cost-functions-for-clarity-native-functions/12386).

> Any modification of cost-limits is a consensus-breaking change. There seems to be broad community support for changing the cost-limits, the question is exactly how and when they go into effect. A previous proposal suggested using a voting contract to determine the block height at which a network-upgrade, described in detail in [this Github discussion](https://github.com/blockstack/stacks-blockchain/discussions/2845). Unfortunately, this path would take at least 4 months in the best-case scenario.

Author: Done

This SIP posits that the ongoing network congestion warrants a more expedient route to change the cost-limits, one that does not rely on an on-chain voting contract.

# Specification

## Assumptions

> This SIP is a method of last resort, considering the circumstances an exception is justified. All future network upgrades should use the voting contract (if appropriate); all hard-forks must follow the process described in SIP-000.

Author: Done

## Proposal

> The Stacks Foundation or the governance group should choose a Bitcoin block height for the network upgrade. The block number should be at least 3 calendar weeks out from when this SIP transitions into “Accepted” state, so as to provide sufficient heads up to node operators.

Reviewer 3: "heads up" is idiomatic.  Please say "provide sufficient time for node operators to upgrade."

Miners, developers, Stackers and community members can demonstrate their support for this network upgrade in one of the following two ways:

> * _Send a contract-call transaction to indicate support for the upgrade_: The contract would aggregate the STX balance on the tx-sender account. It could additionally call into the PoX contract to separately aggregate the total Stacked STX amount. See the Appendix for an initial draft of such a contract.

Reviewer 3: Please label the specific appendix.

* _Send a BTC transaction to indicate support_: Many large STX holders are on multi-sig BTC wallets unable to issue contract-calls. Such wallets could instead send a pure BTC transaction to indicate their support for the vote; it would be a no-op for the Stacks chain, but can be verified off-chain see [Bitcoin support indication](#bitcoin-support-indication) for details.

> The SIP will be considered Recommended if wallets indicating support for the upgrade (through either mechanism) add up to > 10% of circulating supply of STX (approx 120M).

Author: Done

In terms of how these cost-limits would actually be applied, this SIP proposes the following:

> * Add new functionality to stacks-blockchain that uses the current cost-limits by default and uses new cost-limits if the burn block height exceeds a configurable parameter (could be a compile time configuration to avoid runtime issues)

Reviewer 3: "Add new functionality to the Stacks blockchain".

Author: Done

* In the subsequent release, remove all usage of the old cost-limits and just use the new cost-limits by default

### Default Cost Functions

Based on results from the [clarity-benchmarking](https://github.com/blockstack/clarity-benchmarking) project, we propose
new default cost functions. The new costs are supplied in the form of a new Clarity smart contract in [Appendix A](#appendix-a).

### Block Limit Changes

In addition to the runtime cost changes, we propose increasing the
block limits for MARF reads and writes. Based on the expected performance
improvements in the implementation of the MARF (see [issue #2869](https://github.com/blockstack/stacks-blockchain/issues/2869))
we propose doubling the current limits on blocks:

```rust
pub const BLOCK_LIMIT_MAINNET: ExecutionCost = ExecutionCost {
    write_length: 15_000_000, // roughly 15 mb
    write_count: 15_500,
    read_length: 100_000_000,
    read_count: 15_500,
    runtime: 5_000_000_000,
};
```

### Changes to Static vs. Dynamic Tabulation of Costs

> The cost assessment in Clarity for most data-handling functions (e.g.,

Reviewer 3: Per @kantai, this also applies to NFTs and will be done in the `next-costs` branch of `stacks-blockchain`.

`map-get?`) use the static cost of the fetch rather than the dynamic

> cost. For more information, see [issue #2864](https://github.com/blockstack/stacks-blockchain/issues/2864)

Reviewer 3: Can you expand on the problem in this text?  It's fine to link to github issues for supplementary information, but the SIP should be self-contained.  A reader should be able to read this SIP, and only this SIP, to determine what is meant by "static cost of the fetch rather than the dynamic cost."

in the `stacks-blockchain` repository.

There were two motivating reasons for doing this:

* It makes static analysis of costs easier, because the cost assessed
  at runtime would always use the declared size of the map entry.
* It allows the cost to be assessed before running the operation.

However, it's pretty easy to argue that these reasons aren't all that
essential. For (1), static analysis is always going to overestimate
anyways, so why not allow the real cost to match the actual runtime
overhead more closely. For (2), allowing a single "speculative"
evaluation before aborting a block due to cost overflow is not
particularly burdensome to the network: the maximum size of an
overread is a single Clarity value, 2MB.

> The benefit of using dynamic costs, however, could be great. Many

Reviewer 3: This is a tautology -- benefits are always great.  "The benefit of using dynamic costs, however, is that many contracts that store potentially long lists to their data space would only be billed for the amount of space the list occupies, instead of its maximum size."

contracts use patterns where potentially long lists are stored in data
maps, but in practice the stored lists are relatively short.

Because of this, we propose to use a dynamic cost for these assessments.

### Bitcoin Support Indication

For wallets that wish to submit a supporting vote for the activation of this SIP via a BTC transaction, they can submit a transaction in the following form:

The first input to the Bitcoin operation _must_ consume a UTXO that is

> a standard P2MS over P2SH output _or_ a standard P2PKH output. This first

Reviewer 3: What's "P2MS"?

input will be used to verify that the sender of the Bitcoin operation corresponds
to a specific Stacks address.

The Bitcoin transaction will include an `OP_RETURN` output for the first Bitcoin output:

```
            0      2  3       9
            |------|--|-------|
              XM    A  SIP-12
```

That is, the `OP_RETURN` output will be a 9-byte field with content equal to `XMASIP-12` (ascii encoded).

# Activation

The SIP will be considered Active once:

* A new release of stacks-blockchain is available with the updated cost-limits and a mechanism to use the new cost-limits beyond a pre-determined Bitcoin block height

> * This new release is deployed by independent miners, as determined by the continued operation of the Stacks blockchain beyond the Bitcoin block height selected for the network-upgrade. 

Reviewer 3: All miners have to switch to the new cost limits at once in order for the chain to not split apart.  The way systems like Bitcoin handle this is to execute a miner vote -- if some threshold of miners all set a particular bit pattern in a scratch area in the block (e.g. the coinbase transaction, which only a miner can write), and do so for a given amount of time, then it's taken as evidence that the miners are running the software with the proposed changes.  I think we could do that with Stacks as well -- miners would set a particular bit pattern in their coinbase transactions to signal that they're in support of this SIP.  If there exists a complete reward cycle where over 90% of blocks mined have set this bit pattern, then we can say that miners have upgraded and the SIP is safe to activate.

> 

Reviewer 3: The Activation section does not describe how long the voting process is allowed to take.  Please put a reasonable deadline on it (i.e. no more than 3 months of Bitcoin blocks).  Like, if we don't get enough support by then, then this SIP should fail to activate.

The Activation section also does not consider miners as a stakeholder, even though they would be affected by the upgrade.  Please describe a protocol by which they can signal support (or rejection), and update the Activation section to require that both the Stackers and miners have signed off on the upgrade.

        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_nft_transfer (n uint))
    {
        runtime: (linear n u1 u787),
        write_length: u1,
        write_count: u1,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_nft_owner (n uint))
    {
        runtime: (linear n u1 u680),
        write_length: u0,
        write_count: u0,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_ft_get_supply (n uint))
    {
        runtime: u474,
        write_length: u0,
        write_count: u0,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_ft_burn (n uint))
    {
        runtime: u599,
        write_length: u1,
        write_count: u2,
        read_count: u2,
        read_length: u1
    })

(define-read-only (cost_nft_burn (n uint))
    {
        runtime: (linear n u1 u644),
        write_length: u1,
        write_count: u1,
        read_count: u1,
        read_length: u1
    })

(define-read-only (poison_microblock (n uint))
    {
        runtime: u29374,
        write_length: u1,
        write_count: u1,
        read_count: u1,
        read_length: u1
    })

> ```

Reviewer 3: I know it was mentioned above, but for the reader's benefit, please briefly describe here how these numbers were arrived at, and describe how the reader can reproduce them.

## Appendix B

> Initial sketch of a simply Clarity contract to tally up support for the network-upgrade from STX holders. This version does not call into the PoX contract to separately aggregate Stacked STX -- the contract can be modified to do that or that can be    computed offline.

Reviewer 3: I think there needs to be testnet-deployed version of this contract before it can be moved to Recommended status.  We need to see that this Stacker-vote process works, and we need to be able to test it.

Reviewer 3: Once this is done, please update the SIP with the transaction ID and address of the testnet contract.

# Preamble

SIP Number: 012

Title: Burn Height Selection for a Network Upgrade to Introduce New Cost-Limits

Authors:

> * 0xAsteria <asteria@syvita.org>

Reviewer 2: ```suggestion
* Asteria <asteria@syvita.org>
```

* Aaron Blankstein <aaron@hiro.so>
* Diwaker Gupta <diwaker@hiro.so>
* Jason Lau <jason@okcoin.com>
* Ludovic Galabru <ludo@hiro.so>
* Trevor Owens <trevor@stacks.ac>
* Xan Ditkoff <xan@daemontechnologies.co>

Consideration: Governance, Technical

Type: Consensus

Status: Draft

Created: 2021-10-08

License: BSD 2-Clause

Sign-off:

Discussions-To: https://github.com/stacksgov/sips

# Abstract

The current Clarity cost limits were set very conservatively in Stacks 2.0:
transactions with contract-calls frequently exceed these limits, which
negatively affects transaction throughput. This SIP proposes an update to these
cost-limits via a network upgrade and further, that the network upgrade be
executed at a block height chosen by an off-chain process described in this SIP.

# License and Copyright

This SIP is made available under the terms of the Creative Commons CC0 1.0
Universal license, available at

> https://creativecommons.org/publicdomain/zero/1.0/ This SIP’s copyright is held

Reviewer 2: ```suggestion
https://creativecommons.org/publicdomain/zero/1.0/ 

This SIP’s copyright is held
```

description of all cost dimensions) found that roughly 72% of those blocks hit
the `runtime` limit and ~27% hit the `read_count` limit. Another analysis of all
transactions rejected due to cost limits found that 90% of those transactions
exceeded the `runtime` limit. So the `runtime` limits are the primary bottleneck
right now.

In the last few months, the
[clarity-benchmarking](https://github.com/blockstack/clarity-benchmarking)
project has done rigorous benchmarking on contemporary hardware to come up with
more accurate cost-limits, with a focus on the `runtime` limits. The updated
cost-limits are described in detail in [this forum
post](https://forum.stacks.org/t/more-accurate-cost-functions-for-clarity-native-functions/12386).

Any modification of cost-limits is a consensus-breaking change. We believe that
there is broad community support for changing the cost-limits, the question is exactly
how and when they go into effect. A previous proposal suggested using a voting
contract to determine the block height at which a network-upgrade, described in
detail in [this Github
discussion](https://github.com/blockstack/stacks-blockchain/discussions/2845).

> Unfortunately, this path would take at least 4 months in the best-case scenario.

Reviewer 4: Is there a description how the 4 months are estimated?

This SIP posits that the ongoing network congestion warrants a more expedient

> route to change the cost-limits, one that does not rely on an on-chain voting

Reviewer 4: What is the difference between the proposed voting contract in appendix #B and the exiting one?

contract. Equally of note, we consider this SIP as a method of last resort but

> that considering the circumstances an exception is justified. All future network

Reviewer 4: "All future network upgrade..." This was true already before SIP-012 proposal. This SIP could define a list of circumstances when a similar process of SIP-012 could be applied. 

To me, it is not clear whether the sip defines a general process to be applied under exceptional circumstances (how to set the block height, etc) or whether the SIP defines a one-off process (because applying the same values of Appendix A would not change anything after the values have be used as new cost limits).
 
To avoid confusion, we should only use "network upgrade", or introduce `hard-fork` as an equivalent term. (Maybe there is already in one of the sips).

Reviewer 3: SIP-011 is an attempt to clarify when breaking changes in the consensus layer are allowed to occur, but it's still a draft.

I'm of the opinion that this SIP should only define a one-off process, because there's already a general process being developed.  There can be a subsequent SIP that re-proposes this process as a general process if it turns out to work really well.

upgrades should use the voting contract (if appropriate); all hard-forks must
follow the processes described in SIP-000 and SIP-011.

# Specification

> ## Assumptions

Reviewer 2: ```suggestion

```

## Proposal

The Stacks Foundation or the governance group should choose a Bitcoin block
height for the network upgrade. The block number should be at least two calendar
weeks from when this SIP transitions into “Accepted” state, so as to provide
sufficient time for node operators to upgrade.

Miners, developers, Stackers and community members can demonstrate their support
for this network upgrade in one of the following two ways:

* _Send a contract-call transaction to indicate support for the upgrade_: The
  contract would aggregate the STX balance on the `tx-sender` account. It could
  additionally call into the PoX contract to separately aggregate the total

>   Stacked STX amount. See the Appendix for an initial draft of such a contract.

Reviewer 4: What is the role of the initial draft here? Does it have any meaning for the SIP or the activation process? It suggests a certain solution without implying any. I mean the final implementation could be different, couldn't it?

* _Send a BTC transaction to indicate support_: Many large STX holders are on
  multi-sig BTC wallets unable to issue contract-calls. Such wallets could
  instead send a pure BTC transaction to indicate their support for the vote. It
  would be a no-op for the Stacks chain, but can be verified off-chain -- see
  [Bitcoin support indication](#bitcoin-support-indication) for details.

> The SIP will be considered Recommended if wallets indicating support for the

Reviewer 2: ```suggestion
The SIP will be considered _Recommended_ if wallets indicating support for the
```

* Add new functionality to Stacks blockchain that uses the current cost-limits
  by default and uses new cost-limits if the burn block height exceeds a
  configurable parameter (could be a compile time configuration to avoid runtime
  issues)
* Once a Bitcoin block number has been determined, ship a new stacks-blockchain
  release at least one week before to give miners and node operators time to
  upgrade before the upgrade block height is reached

### Default Cost Functions

Based on results from the
[clarity-benchmarking](https://github.com/blockstack/clarity-benchmarking)
project, we propose new default cost functions. The new costs are supplied in
the form of a new Clarity smart contract in [Appendix A](#appendix-a).

### Block Limit Changes

In addition to the runtime cost changes, we propose increasing the block limits
for MARF reads and writes. Based on the expected performance improvements in the
implementation of the MARF (see [issue
#2869](https://github.com/blockstack/stacks-blockchain/issues/2869)) we propose
doubling the current limits on blocks:

```rust
pub const BLOCK_LIMIT_MAINNET: ExecutionCost = ExecutionCost {
    write_length: 15_000_000, // roughly 15 mb

>     write_count: 15_500,

Reviewer 2: ```suggestion
    write_count: 15_000,
```

    read_length: 100_000_000,

>     read_count: 15_500,

Reviewer 2: ```suggestion
    read_count: 15_000,
```

    runtime: 5_000_000_000,
};
```

### Changes to Static vs. Dynamic Tabulation of Costs

> The cost assessment in Clarity for most data-handling functions (e.g.,

Reviewer 2: ```suggestion
The cost assessment in Clarity for most functions that interact with the contract's data space (e.g.,
```

`map-get?`) use the static cost of the fetch rather than the dynamic cost. For
more information, see [issue
#2864](https://github.com/blockstack/stacks-blockchain/issues/2864) in the
`stacks-blockchain` repository.

There were two motivating reasons for doing this:

* It makes static analysis of costs easier, because the cost assessed at runtime
  would always use the declared size of the map entry.
* It allows the cost to be assessed before running the operation.

However, it's pretty easy to argue that these reasons aren't all that essential.
For (1), static analysis is always going to overestimate anyways, so why not
allow the real cost to match the actual runtime overhead more closely. For (2),
allowing a single "speculative" evaluation before aborting a block due to cost
overflow is not particularly burdensome to the network: the maximum size of an

> overread is a single Clarity value, 2MB.

Reviewer 2: ```suggestion
overread is a single Clarity value, which takes only 2MB.
```

The benefit of using dynamic costs, however, could be great. Many contracts use
patterns where potentially long lists are stored in data maps, but in practice
the stored lists are relatively short.

Because of this, we propose to use a dynamic cost for these assessments.

### Bitcoin Support Indication

For wallets that wish to submit a supporting vote for the activation of this SIP
via a BTC transaction, they can submit a transaction in the following form:

The first input to the Bitcoin operation _must_ consume a UTXO that is a
standard P2MS over P2SH output _or_ a standard P2PKH output. This first input
will be used to verify that the sender of the Bitcoin operation corresponds to a
specific Stacks address.

The Bitcoin transaction will include an `OP_RETURN` output for the first Bitcoin
output:

```
            0      2  3       9
            |------|--|-------|
              XM    A  SIP-12
```

That is, the `OP_RETURN` output will be a 9-byte field with content equal to

> `XMASIP-12` (ascii encoded).

Reviewer 2: ```suggestion
`XMASIP-12` (ascii encoded to `584d415349502d3132` in hex format).
```

(define-read-only (cost_default_to (n uint))
    (runtime u255))

(define-read-only (cost_unwrap_ret (n uint))
    (runtime u330))

(define-read-only (cost_unwrap_err_or_ret (n uint))
    (runtime u319))

(define-read-only (cost_is_okay (n uint))
    (runtime u275))

(define-read-only (cost_is_none (n uint))
    (runtime u229))

(define-read-only (cost_is_err (n uint))
    (runtime u268))

(define-read-only (cost_is_some (n uint))
    (runtime u217))

(define-read-only (cost_unwrap (n uint))
    (runtime u281))

(define-read-only (cost_unwrap_err (n uint))
    (runtime u273))

(define-read-only (cost_try_ret (n uint))
    (runtime u275))

(define-read-only (cost_match (n uint))
    (runtime u316))

(define-read-only (cost_or (n uint))
    (runtime (linear n u3 u147)))

(define-read-only (cost_and (n uint))
    (runtime (linear n u3 u146)))

(define-read-only (cost_append (n uint))
    (runtime (linear n u1 u1024)))

(define-read-only (cost_concat (n uint))
    (runtime (linear n u1 u1004)))

(define-read-only (cost_as_max_len (n uint))
    (runtime u482))

(define-read-only (cost_contract_call (n uint))
    (runtime u154))

(define-read-only (cost_contract_of (n uint))

>     (runtime u13391))

Reviewer 4: There is an open issue on the benchmark repo suggesting to set this value to less than 1000:
https://github.com/blockstack/clarity-benchmarking/issues/22

Reviewer 5: Yes -- I think both of those issues could change the final proposal here. They'll need to be addressed first via the clarity-benchmarking repository (and it's output) before updating here.

        write_count: u0,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_ft_burn (n uint))
    {
        runtime: u599,
        write_length: u1,
        write_count: u2,
        read_count: u2,
        read_length: u1
    })

(define-read-only (cost_nft_burn (n uint))
    {
        runtime: (linear n u1 u644),
        write_length: u1,
        write_count: u1,
        read_count: u1,
        read_length: u1
    })

(define-read-only (poison_microblock (n uint))
    {
        runtime: u29374,
        write_length: u1,
        write_count: u1,
        read_count: u1,
        read_length: u1
    })
```
## Appendix B

Initial sketch of a simply Clarity contract to tally up support for the
network-upgrade from STX holders. This version does not call into the PoX
contract to separately aggregate Stacked STX -- the contract can be modified to
do that or that can be    computed offline.

```lisp
(define-data-var signaled-support uint u0)

(define-map known-tx-senders principal bool)

(define-public (in-favor-of-cost-upgrade)
    (let ((existing-entry (map-get? known-tx-senders tx-sender)))
    (asserts! (is-none existing-entry) (err u0))

>     (var-set signaled-support (+ (stx-get-balance tx-sender) (var-get signaled-support)))

Reviewer 4: plus stacked amount: 
`(match (contract-call? 'SP000000000000000000002Q6VF78.pox get-stacker-info tx-sender) info (get amount-ustx info) u0)`

        read_count: u2,
        read_length: u1
    })

(define-read-only (cost_ft_balance (n uint))
    {
        runtime: u543,
        write_length: u0,
        write_count: u0,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_nft_mint (n uint))
    {
        runtime: (linear n u1 u724),
        write_length: u1,
        write_count: u1,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_nft_transfer (n uint))
    {
        runtime: (linear n u1 u787),
        write_length: u1,
        write_count: u1,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_nft_owner (n uint))
    {
        runtime: (linear n u1 u680),
        write_length: u0,
        write_count: u0,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_ft_get_supply (n uint))
    {
        runtime: u474,
        write_length: u0,
        write_count: u0,
        read_count: u1,
        read_length: u1
    })

(define-read-only (cost_ft_burn (n uint))
    {
        runtime: u599,
        write_length: u1,
        write_count: u2,
        read_count: u2,
        read_length: u1

>     })

Reviewer 6: Why is there a difference in runtime between ft_burn and ft_transfer? Why wouldn't I choose to ft_transfer to the burn address for less runtime?

Reviewer 6: I understand that these numbers were arrived at with statistical analysis so they are approximations and will variate accordingly. However, knowing that `ft_burn` will be considered as having higher runtime then it is only a disadvantage to use it. Unless there's something I'm missing, they should just have the same runtime.

Same idea applies to the comment below.

(define-read-only (cost_nft_burn (n uint))
    {
        runtime: (linear n u1 u644),
        write_length: u1,
        write_count: u1,
        read_count: u1,
        read_length: u1
    })

> 

Reviewer 6: Also here, the variable is the NFT Type (basing this on sip-006). Why is runtime different?

the goal of this proposal is that this should correspond to 30 seconds
or `3e10` nanoseconds. So, to convert the `runtime_ns` functions into
runtimes for the Stacks network, we have the simple conversion:

```
runtime_stacks = runtime_ns * 5e9 / 3e10ns
```

For running the benchmarks and analysis in the `clarity-benchmarking`
repository, see the [`README.md`](https://github.com/blockstack/clarity-benchmarking/blob/master/README.md)
file in that repository.

## Appendix B

Initial sketch of a simply Clarity contract to tally up support for the
network-upgrade from STX holders. This version does not call into the PoX
contract to separately aggregate Stacked STX -- the contract can be modified to
do that or that can be    computed offline.

```lisp
(define-data-var signaled-support uint u0)

(define-map known-tx-senders principal bool)

(define-public (in-favor-of-cost-upgrade)
    (let ((existing-entry (map-get? known-tx-senders tx-sender)))
    (asserts! (is-none existing-entry) (err u0))
    (var-set signaled-support (+ (stx-get-balance tx-sender) (var-get signaled-support)))
    (map-set known-tx-senders tx-sender true)

>     (ok true))

Reviewer 6: What stops me from transferring funds to another account and then vote twice?

Shouldn't the support amount use the stacked amount only and for the duration of a cycle?

transactions with contract-calls frequently exceed these limits, which
negatively affects transaction throughput. This SIP proposes an update to these
cost-limits via a network upgrade and further, that the network upgrade be
executed at a block height chosen by an off-chain process described in this SIP.

# License and Copyright

This SIP is made available under the terms of the Creative Commons CC0 1.0
Universal license, available at
https://creativecommons.org/publicdomain/zero/1.0/ This SIP’s copyright is held
by the Stacks Open Internet Foundation.

# Introduction

The current Clarity cost limits were set very conservatively in Stacks 2.0.
Since the mainnet launch on 2021-01-14, traffic on the network has grown
steadily. And in recent months we have seen network congestion adversely
impacting user experience on multiple occassions: valid transactions are not
getting processed because they exceed one or more of the block limits, and are
therefore not included by miners.

Here are some relevant results from a [recent
analysis](https://github.com/blockstack/stacks-blockchain/discussions/2883) of
block data:

> * Average number of "events" per block: 34.9. This is lower than the demands of

Reviewer 3: If you're going to report an average, then please also report a standard deviation.  Also, please explicitly state what an "event" is -- is this a Clarity call? A `(print )` statement? An asset transfer? Something else?

Author: Removed

  these addresses.
* The transaction's scriptSig must be signed by either the Stacker's PoX reward
  address's public key(s), or the public key(s) of their Stacks address (the
  option is provided here because not all Stackers have access to their PoX
  addresses). In either case, the vote commits the Stacker's
  most-recently-locked STX to "yes" or "no" if the Stacker had some STX locked
  in the past two reward cycles at the time of the vote.

Nodes that run Stacks 2.05 must put `0x05` in the memo field. Block-commit
transactions that do not have `0x05` will be considered invalid. The purpose of
this change is to ensure that in the unlikely event some miners didn't know
about this SIP, they will quickly find out because their blocks will never be
confirmed.

The SIP will be considered Recommended if the vote to activate Stacks 2.05
passes. This requires:

* 2/3 of all votes passed are "yes", weighted by the STX they represent
* At least 60 million STX are represented by the "yes" votes. This is 2x the
  largest Stacker right now.

The rationale for this voting procedure is that it simultaneously gives the

> community a way to veto the SIP while also accommodating a low-ish turnout. The

Reviewer 3: So, my comment text wasn't meant to be copy/pasted directly ;)  "low-ish" isn't an appropriate vernacular for a standards-track document.  Just use "somewhat low" or the like.

> problem with any blockchain-based voting systems in the past is that unless

Reviewer 3: You will want citations for this.  The ones I was thinking of are Bitshares and the Ethereum carbon vote for the DAO hard fork.  I'm sure there are others (maybe we can cite EOS?)

# Preamble

SIP Number: 012

Title: Burn Height Selection for a Network Upgrade to Introduce New Cost-Limits

Authors:
* Asteria <asteria@syvita.org>
* Aaron Blankstein <aaron@hiro.so>
* Diwaker Gupta <diwaker@hiro.so>

> * Hank Stoever <hank@oby.io>

Reviewer 7: ```suggestion
* Hank Stoever <hank@stackerlabs.co>
```

  is `00000000000000000000000000000000000000ff`.  On mainnet, this is address
`111111111111111111112kmzDG2`.

Note that these are similar addresses to the PoX burn address, but they all
differ from one another in their checksums.

Vote tallying is performed by examining how many STX the sender has recently
Stacked.  The Bitcoin transaction identifies the amount of STX in the `.pox`
smart contract's data space.  So, by examining the UTXOs for these two Bitcoin
addresses, anyone with a full copy of the Stacks chain state as of the voting
deadline will be able to calculate how many recently-Stacked STX have signaled
"yes" or "no" support for this SIP.

To match the Bitcoin transaction to the Stacker's state in the `.pox` contract,
the `scriptSig` of the first transaction input must be signed by either the user's PoX reward
address's public key(s), or the public key(s) of the Stacks address that Stacked
the tokens (the option is provided here because not all Stackers have access to their PoX
addresses). In either case, the vote commits the Stacker's
most-recently-locked STX to "yes" or "no" if the Stacker had some STX locked

> in the past two reward cycles as of the vote deadline block.

Reviewer 3: > @friedger What does this mean for stackers that delegated their stacking?

Excellent question from @friedger over on https://github.com/stacksgov/sips/pull/43.  The answer is that in this design, the pool operator would be voting on behalf of the users that delegated to the pool.  It would be up to the pool operator to figure out which way to vote on behalf of his/her users.

I'm open to making it so that individual users could vote by sending Bitcoin transactions instead, but I'm not sure what the best way to do this would be.  A Bitcoin transaction isn't cheap, so I'm not sure it's something that users would want to do.

I'm also open to making it so that if users disagreed with the pool operator, they would have the option of de-voting their STX by revoking delegation prior to the vote deadline block.  Then, the pool's total STX voted would be the difference between the number of STX Stacked and the number of STX that are locked but no longer delegated.

Maybe we could make it so that if the user de-delegates their STX, they would have the option of sending a Bitcoin transaction to vote on their own?

in bytes of the consensus serialization of the supplied `key`
argument plus the length in bytes of the consensus serialization of
a `none` Clarity value.

The memory usage of this function should be this same `x` value. The
memory usage of `map-delete` remains in effect until the end of the
transaction (data operations remain in memory during the whole
transaction to enable rollbacks and post-conditions).

#### `(concat list-a list-b)`

The `x` input to the `concat` cost function should be the length of
`list-a` plus the length of `list-b`.

### New Default Cost Functions

Based on results from the
[clarity-benchmarking](https://github.com/blockstack/clarity-benchmarking)
project, this SIP proposes new default cost functions. The new costs are supplied in
the form of a new Clarity smart contract in [Appendix A](#appendix-a).

This could have been carried out through a SIP-006 cost voting procedure, but
due to the opportunity costs incurred by STX holders foregoing PoX rewards to
carry this vote out, this SIP instead proposes bypassing the SIP-006 voting
procedure in this one instance and instead using this SIP's activation procedure
to install these new functions.

# Activation

> The SIP will be considered Ratified once all of the following are true:

Reviewer 3: Add another criterion: the cost functions have to be finalized.  Thanks @friedger!
